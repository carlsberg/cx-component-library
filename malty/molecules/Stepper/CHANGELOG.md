# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

# 0.2.0 (2024-05-31)

### Bug Fixes

- **DSM-286:** added dataTestId & theme ([#241](https://github.com/CarlsbergGBS/cx-component-library/issues/241)) ([056590a](https://github.com/CarlsbergGBS/cx-component-library/commit/056590a91467e71f389251630050ca02202ca059))
- Run auto fixable lint based on new `eslint` configuration ([#501](https://github.com/CarlsbergGBS/cx-component-library/issues/501)) ([9ee94cc](https://github.com/CarlsbergGBS/cx-component-library/commit/9ee94cc2cd407f717e62d8857b5f20a74e7bbec4))
- Update components to have a function declaration and remove unneeded fragments ([#502](https://github.com/CarlsbergGBS/cx-component-library/issues/502)) ([b470095](https://github.com/CarlsbergGBS/cx-component-library/commit/b4700954f4e1ae9c7e547f4de37ac2b15a9d55ea))

### Features

- add tsconfig files and update package.json files for each component ([3bfc644](https://github.com/CarlsbergGBS/cx-component-library/commit/3bfc644e1cfc9dbb7cf7a0469e25fce055b53240))
- change organization name ([a0aa428](https://github.com/CarlsbergGBS/cx-component-library/commit/a0aa428f43138af5707a3ff4c0e36b7c056f02e6))
- **DSM-192:** added Labels option to stepper ([#186](https://github.com/CarlsbergGBS/cx-component-library/issues/186)) ([688c238](https://github.com/CarlsbergGBS/cx-component-library/commit/688c238caa4070619280376495e13d9d39c831ec))
- **DSM-248:** Headline HTML tags ([#184](https://github.com/CarlsbergGBS/cx-component-library/issues/184)) ([305f5e0](https://github.com/CarlsbergGBS/cx-component-library/commit/305f5e0fa92d5813f4877b338ad827a8c347d452))
- **DSM-362:** snapshots removed & unit test updates ([#307](https://github.com/CarlsbergGBS/cx-component-library/issues/307)) ([78423eb](https://github.com/CarlsbergGBS/cx-component-library/commit/78423eb5cfe04fe8a3a7af1f5af074ac6d35e235))
- **DSM-439:** Change font-family behavior ([#470](https://github.com/CarlsbergGBS/cx-component-library/issues/470)) ([ab6dad4](https://github.com/CarlsbergGBS/cx-component-library/commit/ab6dad414c42c4343de3de6abaa4cd3ff98cfca1))
- **general:** [DSM-953] refactor icons implementation ([#892](https://github.com/CarlsbergGBS/cx-component-library/issues/892)) ([08d259f](https://github.com/CarlsbergGBS/cx-component-library/commit/08d259f476ce97e9db8ac41d5eb4773eabca0d37))
- **global:** [DSM-784] add space props in all components ([#842](https://github.com/CarlsbergGBS/cx-component-library/issues/842)) ([9f03f59](https://github.com/CarlsbergGBS/cx-component-library/commit/9f03f59e7518a99abcea06bacbee989d1ddb0465))
- **global:** migrate from bit to lerna ([a0110c8](https://github.com/CarlsbergGBS/cx-component-library/commit/a0110c8831370dc762c193b17cc593eed381f990))
- **global:** update storybook files to be compliant with v7 ([#970](https://github.com/CarlsbergGBS/cx-component-library/issues/970)) ([c61d4b4](https://github.com/CarlsbergGBS/cx-component-library/commit/c61d4b40a0755becf942ad3f28758a159f8c54e4))
- integrate visual testing ([#925](https://github.com/CarlsbergGBS/cx-component-library/issues/925)) ([35dfc1c](https://github.com/CarlsbergGBS/cx-component-library/commit/35dfc1cce6b76d5d2661d552f9fb25477d3a57f7))
- new icon colors ([#215](https://github.com/CarlsbergGBS/cx-component-library/issues/215)) ([9247d33](https://github.com/CarlsbergGBS/cx-component-library/commit/9247d3370bcde78432033ba8b23837f97e0b0dc0))
- remove empty composition files ([f8173f4](https://github.com/CarlsbergGBS/cx-component-library/commit/f8173f4a2ecbf80bb7b6ffe848c023ae31819c2d))
- Use new `eslint-config-carlsberg` with streamlined way to create components ([#559](https://github.com/CarlsbergGBS/cx-component-library/issues/559)) ([a459332](https://github.com/CarlsbergGBS/cx-component-library/commit/a45933215bc0b523220743f17d988d7f1ad5fce6))
