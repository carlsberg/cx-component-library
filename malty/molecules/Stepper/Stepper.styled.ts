/* eslint-disable no-nested-ternary */
import styled from 'styled-components';

export const StyledStepperContainer = styled.div`
  display: flex;
  align-items: center;
`;

export const StyledStepperCircle = styled.span<{
  active: boolean;
  currentStep: boolean;
}>`
  display: flex;
  justify-content: center;
  width: ${({ theme, active, currentStep }) =>
    currentStep || !active
      ? parseInt(theme.sizes.s.value.replace('px', ''), 10) - 4
      : theme.sizes.s.value.replace('px', '')}px;
  height: ${({ theme, active, currentStep }) =>
    currentStep || !active
      ? parseInt(theme.sizes.s.value.replace('px', ''), 10) - 4
      : theme.sizes.s.value.replace('px', '')}px;
  border: ${({ theme, active, currentStep }) =>
    currentStep
      ? `solid 2px ${theme.colors.colours.default['digital-black'].value}`
      : active
      ? '0'
      : `solid 2px ${theme.colors.colours.support[60].value}`};
  border-radius: 50%;
  align-items: center;
`;

export const StyledStepperNumber = styled.span<{
  active: boolean;
}>`
  font-size: ${({ theme }) => theme.typography.desktop.text.micro_default['font-size'].value};
  font-weight: bold;
  display: flex;
  justify-content: center;
  font-family: ${({ theme }) => theme.typography.desktop.text.micro_default['font-family'].value};
  color: ${({ theme, active }) =>
    active ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[60].value};
`;

export const StyledStepperLine = styled.span<{
  active: boolean;
}>`
  display: inline-block;
  flex: 1;
  height: 4px;
  background-color: ${({ theme, active }) =>
    active ? theme.colors.colours.default['digital-black'].value : theme.colors.colours.support[60].value};
`;
