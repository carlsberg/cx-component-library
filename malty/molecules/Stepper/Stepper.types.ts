import { SpaceProps } from '@carlsberggbs/malty.utils.space';

export interface StepperProps extends SpaceProps {
  steps: number | StepsType[];
  currentStep: number;
  isMultiStep?: boolean;
  dataQaId?: string;
}

export interface StepsType {
  label: string;
  key: number;
}
