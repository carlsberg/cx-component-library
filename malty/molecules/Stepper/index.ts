export { Stepper } from './Stepper';
export type { StepperProps, StepsType } from './Stepper.types';
