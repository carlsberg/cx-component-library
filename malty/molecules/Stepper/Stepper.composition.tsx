import { Stepper } from './Stepper';
import React from 'react';

export const BasicCodeInput = () => <Stepper steps={5} currentStep={2} />;
