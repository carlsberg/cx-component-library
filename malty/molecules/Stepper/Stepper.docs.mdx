---
labels: ['react', 'typescript', 'stepper']
description: 'A Stepper component.'
---

import { Stepper } from './Stepper';

Stepper example:

An simple Stepper component

Live Playground:

```js live
<Stepper steps={5} currentStep={2} />
```
