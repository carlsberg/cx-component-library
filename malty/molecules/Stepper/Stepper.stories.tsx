import { Meta, Story } from '@storybook/react';
import React from 'react';
import { Stepper } from './Stepper';
import { StepperProps } from './Stepper.types';

export default {
  title: 'Molecules/Stepper',
  component: Stepper,
  parameters: {
    importObject: 'Stepper',
    importPath: '@carlsberggroup/malty.molecules.stepper'
  },
  argTypes: {
    steps: { control: 'text' },
    currentStep: { control: 'text' },
    isMultiStep: { control: 'boolean' }
  }
} as Meta;

const Template: Story<StepperProps> = ({ steps, currentStep, isMultiStep }: StepperProps) => (
  <Stepper steps={steps} currentStep={currentStep} isMultiStep={isMultiStep} />
);

export const Main = Template.bind({});
Main.args = {
  steps: 5,
  currentStep: 2
};
