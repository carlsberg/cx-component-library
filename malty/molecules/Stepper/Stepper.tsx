/* eslint-disable no-plusplus */
/* eslint-disable no-loop-func */
import { Icon, IconColor, IconName, IconSize } from '@carlsberggroup/malty.atoms.icon';
import { globalTheme as defaultTheme } from '@carlsberggroup/malty.theme.malty-theme-provider';
import React, { useContext, useEffect, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { StyledStepperCircle, StyledStepperContainer, StyledStepperLine, StyledStepperNumber } from './Stepper.styled';
import { StepperProps } from './Stepper.types';

export const Stepper = ({ steps, currentStep, isMultiStep }: StepperProps) => {
  const theme = useContext(ThemeContext) || defaultTheme;
  const [elHtml, setElHtml] = useState([<></>]);

  useEffect(() => {
    for (let step = 1; step <= steps; step++) {
      setElHtml((prevState: Array<JSX.Element>) => [
        ...prevState,
        <>
          <StyledStepperCircle
            key={`progressStep_circle_${step}`}
            active={currentStep >= step}
            currentStep={currentStep === step}
            theme={theme}
          >
            <>
              {currentStep > step && (
                <Icon name={IconName.ItemCheckFilled} size={IconSize.Small} color={IconColor.Primary} />
              )}
              {currentStep <= step && !isMultiStep && (
                <StyledStepperNumber theme={theme} active={currentStep >= step}>
                  {step}
                </StyledStepperNumber>
              )}
            </>
          </StyledStepperCircle>
          {step < steps && (
            <StyledStepperLine
              key={`progressStep_line_${step}`}
              theme={theme}
              active={currentStep > step && !isMultiStep}
            />
          )}
        </>
      ]);
    }
    return () => {
      setElHtml([<></>]);
    };
  }, [steps, currentStep]);

  return <StyledStepperContainer theme={theme}>{elHtml}</StyledStepperContainer>;
};
