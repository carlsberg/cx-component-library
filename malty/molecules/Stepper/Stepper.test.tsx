import { render } from '@carlsberggbs/malty.utils.test';
import { screen } from '@testing-library/react';
import React from 'react';
import { Stepper } from './Stepper';

describe('molecule stepper', () => {
  it('renders stepper initials', () => {
    render(<Stepper steps={5} currentStep={2} dataQaId="stepper" />);
    expect(screen.getByTestId('stepper')).toBeDefined();
  });
});
