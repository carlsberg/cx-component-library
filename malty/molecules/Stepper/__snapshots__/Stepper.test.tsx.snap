// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`molecule stepper matches snapshot 1`] = `
<div
  className="sc-dkzDqf hoZXih"
/>
`;
